lua/nvim-navic/init.lua

```diff
@@ -25,6 +25,106 @@ local function request_symbol(for_buf, handler, client_id)
 	)
 end
 
+-- relation of 'other' with repect to 'symbol'
+local function symbol_relation(symbol, other)
+	local s = symbol.scope
+	local o = other.scope
+
+	if
+		o["end"].line < s["start"].line
+		or (o["end"].line == s["start"].line and o["end"].character < s["start"].character)
+	then
+		return "before"
+	end
+
+	if
+		o["start"].line > s["end"].line
+		or (o["start"].line == s["end"].line and o["start"].character > s["end"].character)
+	then
+		return "after"
+	end
+
+	if
+		(
+			o["start"].line < s["start"].line
+			or (o["start"].line == s["start"].line and o["start"].character <= s["start"].character)
+		)
+		and (
+			o["end"].line > s["end"].line
+			or (o["end"].line == s["end"].line and o["end"].character >= s["end"].character)
+		)
+	then
+		return "around"
+	end
+
+	return "within"
+end
+
+-- Construct tree structure based on scope information
+-- Could be inaccurate ?? Not intended to be used like this...
+local function symbolInfo_treemaker(symbols)
+	-- convert location to scope
+	for _, node in ipairs(symbols) do
+		node.scope = node.location.range
+		node.scope["start"].line = node.scope["start"].line + 1
+		node.scope["end"].line = node.scope["end"].line + 1
+		node.location = nil
+	end
+
+	-- sort with repect to node height and location
+	-- nodes closer to root node come before others
+	-- nodes and same level are arranged according to scope
+	table.sort(symbols, function(a, b)
+		local loc = symbol_relation(a, b)
+		if loc == "after" or loc == "within" then
+			return true
+		end
+		return false
+	end)
+
+	-- root node
+	local tree = {
+		scope = {
+			start = {
+				line = -10,
+				character = 0,
+			},
+			["end"] = {
+				line = 2147483640,
+				character = 0,
+			},
+		},
+		children = {},
+	}
+	local stack = {}
+
+	table.insert(tree.children, symbols[1])
+	table.insert(stack, tree)
+
+	-- build tree
+	for i = 2, #symbols, 1 do
+		local prev_chain_relation = symbol_relation(symbols[i], symbols[i - 1])
+		local prev_node_relation = symbol_relation(symbols[i], stack[#stack])
+
+		if prev_chain_relation == "around" then
```

Comment:
Consider adding an explanatory comment.
```suggestion
		if prev_chain_relation == "around" then
			-- the previous symbol contains this symbol; push the previous symbol onto
			-- the stack and make this symbol it's child
```